File: lib/core/text_diff.dart
/// Пословное сравнение сказанного с исправленным вариантом.
///
/// Нужно, чтобы игрок видел правку наглядно: где он сказал не то слово, а
/// где вообще пропустил слово. Списком объяснений это не заменяется — по
/// нему не видно, как должна звучать фраза целиком.
///
/// Сравнение идёт по НОРМАЛИЗОВАННЫМ словам (без пунктуации и регистра):
/// распознавание само расставляет точки и заглавные буквы там, где игрок
/// просто сделал паузу, и считать это правкой нельзя — судья их тоже
/// игнорирует.
library;

enum DiffKind {
  /// Слово сказано верно.
  same,

  /// Слово сказано неправильно — в исправленном варианте его нет.
  wrong,

  /// Слово пропущено — есть в исправленном варианте, но не было сказано.
  missing,
}

class DiffPart {
  final String text;
  final DiffKind kind;

  const DiffPart(this.text, this.kind);

  @override
  String toString() => '${kind.name}:$text';
}

String _normalize(String word) =>
    word.toLowerCase().replaceAll(RegExp(r'''[^\p{L}\p{N}']''', unicode: true), '');

List<String> _tokenize(String text) =>
    text.split(RegExp(r'\s+')).where((w) => w.trim().isNotEmpty).toList();

/// Разбор различий между [spoken] (что распознали) и [corrected] (как
/// должно быть). Возвращает исправленный текст по порядку, помечая каждое
/// слово: сказано верно, сказано неверно или пропущено.
///
/// Если [corrected] пуст — правки нет, весь текст помечается как верный:
/// это не значит «всё правильно», это значит «сравнивать не с чем», и
/// вызывающий код решает, показывать ли блок вообще.
List<DiffPart> diffWords(String spoken, String corrected) {
  final a = _tokenize(spoken);
  final b = _tokenize(corrected);
  if (b.isEmpty) return [for (final w in a) DiffPart(w, DiffKind.same)];

  final na = [for (final w in a) _normalize(w)];
  final nb = [for (final w in b) _normalize(w)];

  // Наибольшая общая подпоследовательность: слова, которые игрок сказал
  // верно и в верном порядке. Всё, что не попало в неё, — либо лишнее/
  // неправильное у игрока, либо пропущенное им.
  final lcs = List.generate(na.length + 1, (_) => List<int>.filled(nb.length + 1, 0));
  for (var i = na.length - 1; i >= 0; i--) {
    for (var j = nb.length - 1; j >= 0; j--) {
      lcs[i][j] = na[i] == nb[j] ? lcs[i + 1][j + 1] + 1 : (lcs[i + 1][j] >= lcs[i][j + 1] ? lcs[i + 1][j] : lcs[i][j + 1]);
    }
  }

  final parts = <DiffPart>[];
  var i = 0;
  var j = 0;
  while (i < na.length && j < nb.length) {
    if (na[i] == nb[j]) {
      parts.add(DiffPart(b[j], DiffKind.same));
      i++;
      j++;
    } else if (lcs[i + 1][j] >= lcs[i][j + 1]) {
      parts.add(DiffPart(a[i], DiffKind.wrong));
      i++;
    } else {
      parts.add(DiffPart(b[j], DiffKind.missing));
      j++;
    }
  }
  while (i < na.length) {
    parts.add(DiffPart(a[i], DiffKind.wrong));
    i++;
  }
  while (j < nb.length) {
    parts.add(DiffPart(b[j], DiffKind.missing));
    j++;
  }
  return parts;
}

File: supabase/functions/_shared/textDiff.ts
/**
 * Пословное сравнение услышанного с правильным вариантом.
 *
 * ЗАЧЕМ ЭТО НА СЕРВЕРЕ. Разметку разбора — что сказано верно, что не так,
 * что пропущено — раньше рисовала сама модель. Дважды подряд она рисовала
 * её неправильно: то помечала сказанное как пропущенное, то объявляла
 * «ошибок нет» на половине фразы. Причина не в плохой модели, а в том, что
 * задача не для неё: сравнить две строки по словам — арифметика, и
 * арифметику надо считать, а не спрашивать.
 *
 * Теперь модель отвечает только за то, что умеет одна она: услышать,
 * перевести и объяснить. Границы кусков считает этот код, и ошибиться в
 * них он не может.
 *
 * Сравнение идёт по НОРМАЛИЗОВАННЫМ словам (без пунктуации и регистра):
 * модель сама расставляет точки и заглавные буквы там, где игрок просто
 * сделал паузу, и считать это ошибкой нельзя.
 *
 * Порт lib/core/text_diff.dart — тот же алгоритм слово в слово, чтобы
 * сервер и клиент не разошлись в том, что считать правкой.
 */

export type DiffKind =
  /** Слово сказано верно. */
  | "same"
  /** Слово сказано неправильно — в правильном варианте его нет. */
  | "wrong"
  /** Слово пропущено — есть в правильном варианте, но сказано не было. */
  | "missing";

export interface DiffPart {
  text: string;
  kind: DiffKind;
}

function normalize(word: string): string {
  return word.toLowerCase().replace(/[^\p{L}\p{N}']/gu, "");
}

function tokenize(text: string): string[] {
  return text.split(/\s+/).filter((w) => w.trim().length > 0);
}

/**
 * Различия между [spoken] (что услышала модель) и [corrected] (как должно
 * быть). Возвращает правильный текст по порядку, помечая каждое слово.
 *
 * Пустой [corrected] — сравнивать не с чем; всё помечается верным, и
 * вызывающий сам решает, показывать ли разбор вообще.
 */
export function diffWords(spoken: string, corrected: string): DiffPart[] {
  const a = tokenize(spoken);
  const b = tokenize(corrected);
  if (b.length === 0) return a.map((text) => ({ text, kind: "same" as const }));

  const na = a.map(normalize);
  const nb = b.map(normalize);

  // Наибольшая общая подпоследовательность: слова, которые игрок сказал
  // верно и в верном порядке. Всё, что не попало в неё, — либо лишнее у
  // игрока, либо пропущенное им.
  const lcs: number[][] = Array.from(
    { length: na.length + 1 },
    () => new Array<number>(nb.length + 1).fill(0),
  );
  for (let i = na.length - 1; i >= 0; i--) {
    for (let j = nb.length - 1; j >= 0; j--) {
      lcs[i][j] = na[i] === nb[j]
        ? lcs[i + 1][j + 1] + 1
        : (lcs[i + 1][j] >= lcs[i][j + 1] ? lcs[i + 1][j] : lcs[i][j + 1]);
    }
  }

  const parts: DiffPart[] = [];
  let i = 0;
  let j = 0;
  while (i < na.length && j < nb.length) {
    if (na[i] === nb[j]) {
      parts.push({ text: b[j], kind: "same" });
      i++;
      j++;
    } else if (lcs[i + 1][j] >= lcs[i][j + 1]) {
      parts.push({ text: a[i], kind: "wrong" });
      i++;
    } else {
      parts.push({ text: b[j], kind: "missing" });
      j++;
    }
  }
  while (i < na.length) parts.push({ text: a[i], kind: "wrong" }), i++;
  while (j < nb.length) parts.push({ text: b[j], kind: "missing" }), j++;
  return parts;
}
